overlord/devicestate/devicestate.go

```diff
@@ -433,7 +435,47 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
 		(ms.newModelSnap.SnapType == "kernel" || ms.newModelSnap.SnapType == "gadget")
 }
 
-func remodelEssentialSnapTasks(ctx context.Context, st *state.State, ms modelSnapsForRemodel, deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
+type remodelOptions struct {
+	localSnapsRequired bool
+}
+
+func (ro *remodelOptions) UpdateWithDeviceContext(st *state.State,
+	localSi *snap.SideInfo, path, snapName, channel string,
+	userID int, snapStateFlags snapstate.Flags,
+	deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
+	logger.Debugf("snap %s track changed", snapName)
+	if ro.localSnapsRequired {
+		if localSi == nil {
+			return nil, fmt.Errorf("no snap file provided for %q (track changed)", snapName)
+		}
+		return snapstateUpdatePathWithDeviceContext(st, localSi, path, snapName,
+			&snapstate.RevisionOptions{Channel: channel},
+			userID, snapStateFlags, deviceCtx, fromChange)
+	}
+	return snapstateUpdateWithDeviceContext(st, snapName,
+		&snapstate.RevisionOptions{Channel: channel},
+		userID, snapStateFlags, deviceCtx, fromChange)
+}
+
+func (ro *remodelOptions) InstallWithDeviceContext(ctx context.Context, st *state.State,
+	localSi *snap.SideInfo, path, snapName, channel string, userID int,
+	snapStateFlags snapstate.Flags, deviceCtx snapstate.DeviceContext,
+	fromChange string) (*state.TaskSet, error) {
+	logger.Debugf("snap %s needs install", snapName)
+	if ro.localSnapsRequired {
+		if localSi == nil {
+			return nil, fmt.Errorf("no snap file provided for %q", snapName)
+		}
+		return snapstateInstallPathWithDeviceContext(st, localSi, path, snapName,
+			&snapstate.RevisionOptions{Channel: channel},
+			userID, snapStateFlags, deviceCtx, fromChange)
+	}
+	return snapstateInstallWithDeviceContext(ctx, st, snapName,
+		&snapstate.RevisionOptions{Channel: channel},
+		userID, snapStateFlags, deviceCtx, fromChange)
+}
+
+func remodelEssentialSnapTasks(ctx context.Context, st *state.State, localSi *snap.SideInfo, path string, ms modelSnapsForRemodel, remodelOpts remodelOptions, deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
```

Comment:
I think we need an options structure for these parameters, there is pretty many. Maybe some of them can go into `remodelOptions`?

Reply:
I have added an ancillary structure named `pathSideInfo` to hold data for local snaps. I don't see that in remodelOptions, which should have only booleans imo.